### PR TITLE
Avoid some defensive copies on readonly structs in System.Net.Quic

### PR DESCRIPTION
**Pull request by rzikm**
This removes defensive copies on readonly structs in S.N.S uncovered by an analyzer from a community member (thanks @hamarb123!).

### CODE REVIEW

**Review by liveans — APPROVED** [review]

**Comment by MihaZupan** [comment]
cc: @hamarb123

**Review by MihaZupan — APPROVED** [review]
Thanks

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs [hunk 1]
```diff
@@ -47,6 +49,8 @@ private static bool GetConfigurationCacheEnabled()
         public readonly List<SslApplicationProtocol> ApplicationProtocols;
         public readonly QUIC_ALLOWED_CIPHER_SUITE_FLAGS AllowedCipherSuites;
 
+        private readonly ReadOnlySpan<byte> SettingsSpan => MemoryMarshal.AsBytes(new ReadOnlySpan<QUIC_SETTINGS>(ref Unsafe.AsRef(in Settings)));
```
Comment (hamarb123) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:52]: You should be able to do `MemoryMarshal.AsBytes(new ReadOnlySpan<QUIC_SETTINGS>(in Settings))` if you're on 7+

**Review by hamarb123 — APPROVED** [review]

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs (hunk 1 above)
Comment (hamarb123) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:52]: You should be able to do `MemoryMarshal.AsBytes(new ReadOnlySpan<QUIC_SETTINGS>(in Settings))` if you're on 7+ (the docs lie and say `ref` as opposed to `ref readonly`)

Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:52]:
```
error CS8347: Cannot use a result of 'ReadOnlySpan<QUIC_SETTINGS>.ReadOnlySpan(ref readonly QUIC_SETTINGS)' in this context because it may expose variables referenced by parameter 'reference' outside of their declaration scope
```

It works if I inline it though, good to know.

Comment (hamarb123) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:52]: Ah yes, to make it safe, you need to mark the member as `[UnscopedRef]`

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/msquic.cs [hunk 2]
```diff
@@ -21,6 +21,7 @@ namespace Microsoft.Quic
     internal unsafe partial struct QUIC_BUFFER
     {
         public Span<byte> Span => new(Buffer, (int)Length);
+        public readonly ReadOnlySpan<byte> ReadOnlySpan => new(Buffer, (int)Length);
```
Comment (hamarb123) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/msquic.cs:24]: If Buffer doesn't refer to memory in the struct / same allocation itself (which it doesn't seem to to me), the existing `Span` member could actually just be marked `readonly` instead of adding a new member - if it does refer to memory within the same allocation / struct, then they should both be marked `[UnscopedRef]`.

Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/msquic.cs:24]: It refers to native memory (Buffer is void*).

I was unsure whether to make the Span property itself readonly instead. I don't really like being able to mutate data through readonly QUIC_BUFFER instance (hence explicit ReadOnlySpan property), but fact is that the possibility is not going away and keeping it like this allows the defensive copy problem to pop up elsewhere, so I will probably just make the Span readonly

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/msquic.cs [hunk 3]
```diff
@@ -20,7 +20,7 @@ namespace Microsoft.Quic
 {
     internal unsafe partial struct QUIC_BUFFER
     {
-        public Span<byte> Span => new(Buffer, (int)Length);
+        public readonly Span<byte> Span => new(Buffer, (int)Length);
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/msquic.cs:23]: I don't remember exactly, but if this is generated code from MsQuic, it should be fixed there as well.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs [hunk 4]
```diff
@@ -47,6 +49,8 @@ private static bool GetConfigurationCacheEnabled()
         public readonly List<SslApplicationProtocol> ApplicationProtocols;
         public readonly QUIC_ALLOWED_CIPHER_SUITE_FLAGS AllowedCipherSuites;
 
+        [UnscopedRef] private readonly ReadOnlySpan<byte> SettingsSpan => MemoryMarshal.AsBytes(new ReadOnlySpan<QUIC_SETTINGS>(in Settings));
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:52]: What is this saving us interms of actual perf improvement? How many peoply actually understand what `UnscopedRef` and this property code does and why is this important to have?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/msquic.cs (hunk 3 above)
Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/msquic.cs:23]: right, I thought this was added manually by us (since the QUIC_BUFFER definition is in another auto-generated file). I will follow up to make the change in their repo as well.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs (hunk 4 above)
Comment (hamarb123) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:52]: It's about 1400x faster and allocates 2kB less per comparison:

image


```csharp
using System.Runtime.InteropServices;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<Benchmarks>();

[MemoryDiagnoser(false)]
public class Benchmarks
{
	private readonly QUIC_SETTINGS a = default;
	private readonly QUIC_SETTINGS b = default;

	[Benchmark]
	public bool Compare1()
	{
		return a.Equals(b);
	}

	[Benchmark]
	public bool Compare2()
	{
		return MemoryMarshal.AsBytes(new ReadOnlySpan<QUIC_SETTINGS>(in a)).SequenceEqual(MemoryMarshal.AsBytes(new ReadOnlySpan<QUIC_SETTINGS>(in b)));
	}
}
```

Comment (hamarb123) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:52]: The above is the worst case obviously, where the settings are completely equal.

Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:52]: The original code (`Settings.Equals(other.Settings)`) used the default struct equality comparer which:
- boxes second argument
- makes a defensive copy of the first argument (since Equals is not a readonly method)
- potentially uses reflection-based comparison (probably not, since the type does not contain a reference, but I am not 100% sure).

This change eliminates the above and does not require us to keep track of new members.

As for `UnscopedRef`, the previous version had 

```cs
private readonly ReadOnlySpan<byte> SettingsSpan =>
    MemoryMarshal.AsBytes(new ReadOnlySpan<QUIC_SETTINGS>(ref Unsafe.AsRef(in Settings)));
```

`UnscopedRef` allows to get rid of Unsafe.AsRef and make it a bit shorter. The property is private to limit potential exposure to other code.

Comment (hamarb123) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:52]: The `UnscopedRef` is also ideal for correctness, otherwise you might accidentally leak the ref out of the struct and lifetime rules enforcement will fail you since you broke the rules.

Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:52]: Formatting nit
```suggestion
        [UnscopedRef] 
        private readonly ReadOnlySpan<byte> SettingsSpan => MemoryMarshal.AsBytes(new ReadOnlySpan<QUIC_SETTINGS>(in Settings));
```

Comment (hamarb123) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs]: I think there might be some padding, e.g. between `MtuDiscoveryMissingProbeCount` and `DestCidUpdateIdleTimeoutMs` it seems like there'd be 3 bytes of padding - which would make the `SequenceEquals` approach incorrect.

If we check each field manually, we still get very similar performance numbers to the `SequenceEquals` approach:
<details>
<summary>
Code for Compare3
</summary>

```csharp
	[Benchmark]
	public bool Compare3()
	{
		return a.Anonymous1.IsSetFlags == b.Anonymous1.IsSetFlags
			&& a.MaxBytesPerKey == b.MaxBytesPerKey
			&& a.HandshakeIdleTimeoutMs == b.HandshakeIdleTimeoutMs
			&& a.IdleTimeoutMs == b.IdleTimeoutMs
			&& a.MtuDiscoverySearchCompleteTimeoutUs == b.MtuDiscoverySearchCompleteTimeoutUs
			&& a.TlsClientMaxSendBuffer == b.TlsClientMaxSendBuffer
			&& a.TlsServerMaxSendBuffer == b.TlsServerMaxSendBuffer
			&& a.StreamRecvWindowDefault == b.StreamRecvWindowDefault
			&& a.StreamRecvBufferDefault == b.StreamRecvBufferDefault
			&& a.ConnFlowControlWindow == b.ConnFlowControlWindow
			&& a.MaxWorkerQueueDelayUs == b.MaxWorkerQueueDelayUs
			&& a.MaxStatelessOperations == b.MaxStatelessOperations
			&& a.InitialWindowPackets == b.InitialWindowPackets
			&& a.SendIdleTimeoutMs == b.SendIdleTimeoutMs
			&& a.InitialRttMs == b.InitialRttMs
			&& a.MaxAckDelayMs == b.MaxAckDelayMs
			&& a.DisconnectTimeoutMs == b.DisconnectTimeoutMs
			&& a.KeepAliveIntervalMs == b.KeepAliveIntervalMs
			&& a.CongestionControlAlgorithm == b.CongestionControlAlgorithm
			&& a.PeerBidiStreamCount == b.PeerBidiStreamCount
			&& a.PeerUnidiStreamCount == b.PeerUnidiStreamCount
			&& a.MaxBindingStatelessOperations == b.MaxBindingStatelessOperations
			&& a.StatelessOperationExpirationMs == b.StatelessOperationExpirationMs
			&& a.MinimumMtu == b.MinimumMtu
			&& a.MaximumMtu == b.MaximumMtu
			&& a._bitfield == b._bitfield
			&& a.MaxOperationsPerDrain == b.MaxOperationsPerDrain
			&& a.MtuDiscoveryMissingProbeCount == b.MtuDiscoveryMissingProbeCount
			&& a.DestCidUpdateIdleTimeoutMs == b.DestCidUpdateIdleTimeoutMs
			&& a.Anonymous2.Flags == b.Anonymous2.Flags
			&& a.StreamRecvWindowBidiLocalDefault == b.StreamRecvWindowBidiLocalDefault
			&& a.StreamRecvWindowBidiRemoteDefault == b.StreamRecvWindowBidiRemoteDefault
			&& a.StreamRecvWindowUnidiDefault == b.StreamRecvWindowUnidiDefault;
	}
```

</details>

image

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs [hunk 5]
```diff
@@ -98,7 +106,7 @@ public bool Equals(CacheKey other)
 
             return
                 Flags == other.Flags &&
-                Settings.Equals(other.Settings) &&
+                SettingsSpan.SequenceEqual(other.SettingsSpan) &&
```
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:109]: Can you add a test pls?

Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:109]: What sort of test do you have in mind?

Comment (CarnaViire) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:109]: I mean the simplest unit test that equal is still equal :D 
see comment above:

> I think there might be some padding, e.g. between MtuDiscoveryMissingProbeCount and DestCidUpdateIdleTimeoutMs it seems like there'd be 3 bytes of padding - which would make the SequenceEquals approach incorrect.

Comment (hamarb123) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:109]: Specifically, the approach using `SequenceEquals` would be incorrect if the padding bytes are different in each, which should compare to equal, but wouldn't.

Comment (CarnaViire) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs:109]: As discussed with @rzikm offline: we should consider updating MsQuic generator code to include an additional field-by-field equality comparison to QUIC_SETTINGS (i.e. approach from  https://github.com/dotnet/runtime/pull/101133#discussion_r1568624284). Adding it to generator will also prevent losing any newly-added fields to the struct.

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicInteropTests.cs [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+            var settingsSpan = MemoryMarshal.AsBytes(new Span<QUIC_SETTINGS>(ref settings));
+
+            // Fill the memory with 1s,we will try to zero out each member and compare
+            settingsSpan.Fill(0xFF);
+
+            foreach (MemberInfo member in GetMembers<QUIC_SETTINGS>())
+            {
+                // copy and box the instance because reflection methods take a reference type arg
+                object boxed = settings;
```
Comment (CarnaViire) [inline src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicInteropTests.cs:72]: Will this copy the padding filling 1s as well?
Can we somehow construct the test (either this one or a separate one) in a way that will verify that the padding values are _not_ compared?

Idea for a separate test: fill with 1s, use reflection to fill all the fields with 0s, compare with `default` (should be equal despite paddings being 1s)

Comment (rzikm) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicInteropTests.cs:72]: we don't care about padding bytes anymore, I replaced the span-based comparison with listing all individual properties.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ReceiveBuffers.cs [hunk 7]
```diff
@@ -52,7 +52,7 @@ public int CopyFrom(ReadOnlySpan<QUIC_BUFFER> quicBuffers, int totalLength, bool
             int totalCopied = 0;
             for (int i = 0; i < quicBuffers.Length; ++i)
             {
-                Span<byte> quicBuffer = quicBuffers[i].Span;
+                ReadOnlySpan<byte> quicBuffer = quicBuffers[i].Span;
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ReceiveBuffers.cs:55]: This is fine, but what does it fix?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/QUIC_SETTINGS.IEquattable.cs [hunk 8]
```diff
@@ -0,0 +1,91 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+
+namespace Microsoft.Quic
+{
+    internal partial struct QUIC_SETTINGS : System.IEquatable<QUIC_SETTINGS>
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/QUIC_SETTINGS.IEquattable.cs:8]: Why is this needed? Is QUIC_SETTINGS being put into a Dictionary somewhere or something?

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicInteropTests.cs (hunk 6 above)
Comment (CarnaViire) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicInteropTests.cs:72]: Well, you replaced it for a _reason_ which is not currently stated anywhere in the code; so if in future someone (not aware of this discussion) will decide to bring the span-based comparison "back", they won't see any test failures.

It's like when you fix some bug but you still add a test to ensure the bug does not and will not reproduce anymore with your changes. It would be a kinda similar situation if we say "but we fixed the bug, so now we don't care about the previously broken scenario anymore"..?

That being said, this situation with padding feels like a corner case, and I guess it is unlikely that span-based comparison would be reintroduced. So I don't feel so strongly about that. More like trying to be extra cautious.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/QUIC_SETTINGS.IEquattable.cs (hunk 8 above)
Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/QUIC_SETTINGS.IEquattable.cs:8]: QUIC_SETTINGS is used as part of the cache key for MsQuicConfigurationHandle

https://github.com/dotnet/runtime/blob/e12e2fa6cbdd1f4b0c8ad1b1e2d960a480c21703/src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicConfiguration.Cache.cs#L101

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicSafeHandle.cs [hunk 9]
```diff
@@ -88,7 +88,7 @@ internal sealed class MsQuicContextSafeHandle : MsQuicSafeHandle
     /// Holds a weak reference to the managed instance.
     /// It allows delegating MsQuic events to the managed object while it still can be collected and finalized.
     /// </summary>
-    private readonly GCHandle _context;
+    private GCHandle _context;
```
[the remark this answers is not in the document]
Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicSafeHandle.cs:91]: This one should stay, otherwise the `.Free()` is executed on a defensive copy

https://github.com/dotnet/runtime/blob/1c7b419097c07a21e4a6bd8d7bde85942c1c015d/src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/MsQuicSafeHandle.cs#L133

It probably is not a big deal since `ReleaseHandle` runs exactly once, but we still make sure to cleanup dangling pointers.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ReceiveBuffers.cs (hunk 7 above)
Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Internal/ReceiveBuffers.cs:55]: it's a leftover change from a previous version, I will roll it back

File: src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicInteropTests.cs (hunk 6 above)
Comment (rzikm) [reply src/libraries/System.Net.Quic/tests/FunctionalTests/MsQuicInteropTests.cs:72]: > It's like when you fix some bug but you still add a test to ensure the bug does not (and will not) reproduce anymore with your changes.

I see tests more like a way to assert some assumptions about the behavior of the code. If we don't assume any behavior regarding padding bytes in the production (or test infra) code, then I don't see value of asserting such properties in tests. As opposed to when we find a bug and add a test asserting that in *that* scenario the code behaves the correct way.

If somebody decides to reintroduce the span-based approach, then I expect the question "is this safe/correct/sound?" to come up during code review (as it did on this PR).

I can add a comment to the implementation so that the reason for the decision is not lost.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/QUIC_SETTINGS.IEquattable.cs [hunk 10]
```diff
@@ -0,0 +1,95 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+
+namespace Microsoft.Quic
+{
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/QUIC_SETTINGS.IEquattable.cs:7]:
```suggestion
namespace Microsoft.Quic;
```

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/QUIC_SETTINGS.IEquattable.cs [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    internal partial struct QUIC_SETTINGS : System.IEquatable<QUIC_SETTINGS>
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/QUIC_SETTINGS.IEquattable.cs:8]: Theoretically, you could move this into System.Net.Quic/src/System/Net/Quic/Internal/MsQuicExtensions.cs where we already have something similar for `ToString`. Depends, whether you expect the MsQuic generator ask to get closed soonish or rather not...

**Review by ManickaP — APPROVED** [review]
Nits, otherwise LGTM.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/QUIC_SETTINGS.IEquattable.cs (hunk 11 above)
Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Interop/QUIC_SETTINGS.IEquattable.cs:8]: it does not seem like they will implement it soon, so I will move this

**Review by CarnaViire — APPROVED** [review]
LGTM

**Comment by rzikm** [comment]
CI failures are unrelated.
